torch/lib/c10d/TCPStore.hpp

```diff
@@ -36,29 +70,55 @@ class TCPStoreDaemon {
   void getNumKeysHandler(int socket) const;
   void deleteHandler(int socket);
   void waitHandler(int socket);
+  void watchHandler(int socket);
 
   bool checkKeys(const std::vector<std::string>& keys) const;
+  // Helper function to alerts waiting workers, used in setHandler, getHandler
   void wakeupWaitingClients(const std::string& key);
-
-  void initStopSignal();
-  void closeStopSignal();
-
-  std::thread daemonThread_;
+  // Helper function used when the key is changed
+  // used in setHandler, addHandler, getHandler, deleteHandler
+  void sendKeyUpdatesToClients(
+      const std::string& key,
+      const enum WatchResponseType& type,
+      std::vector<uint8_t>& oldData,
+      std::vector<uint8_t>& newData);
   std::unordered_map<std::string, std::vector<uint8_t>> tcpStore_;
-  // From key -> the list of sockets waiting on it
+  // From key -> the list of sockets waiting on the key
   std::unordered_map<std::string, std::vector<int>> waitingSockets_;
   // From socket -> number of keys awaited
   std::unordered_map<int, size_t> keysAwaited_;
+  // From key -> the list of sockets watching the key
+  std::unordered_map<std::string, std::vector<int>> watchedSockets_;
+};
 
-  std::vector<int> sockets_;
-  int storeListenSocket_;
-#ifdef _WIN32
-  const std::chrono::milliseconds checkTimeout_
-      = std::chrono::milliseconds(10);
-  HANDLE ghStopEvent_;
-#else
-  std::vector<int> controlPipeFd_{-1, -1};
-#endif
+// Separate thread that is launched on all instances (including master)
+// Right now only handles callbacks registered from watchKey()
+class TCPStoreWorkerDaemon : public BackgroundThread {
+ public:
+  explicit TCPStoreWorkerDaemon(int listenSocket);
+  // Adds a callback to run key change
+  void addCallback(
+      std::string key,
+      std::function<
+          void(c10::optional<std::string>, c10::optional<std::string>)> cb);
+  std::condition_variable& getCallbackRegistered() {
+    return callbackRegistered;
+  }
+  bool getCallbackRegisteredData() const {
+    return callbackRegisteredData;
+  }
+  void setCallbackRegisteredData(bool condition) {
+    callbackRegisteredData = condition;
+  }
+
+ private:
+  void run();
+  void callbackHandler(int socket);
+  // List of callbacks map each watched key
+  std::unordered_map<std::string, StoreCallbackFunction> keyToCallbacks;
+  std::mutex keyToCallbacksMutex;
+  std::condition_variable callbackRegistered;
+  bool callbackRegisteredData = false;
```

Comment:
nit on naming convention: other part of the code base usually adds a `_` suffix to private variable names.